### PR TITLE
Add badge statuses

### LINKED ISSUE #1222: Implement badge statuses from Anthrocon
Anthrocon asked us to implement badge statuses for their registration last year. This ended up being a really useful feature and I want it for MAG events, including Classic if possible.

The statuses include:
* New (registrations that are simply not complete, usually because they're unpaid)
* Complete (attendees that can check in without further action)
* Invalid (duplicates, mistakes, and badges we want to 'delete')
* Deferred (attendees who cannot check in for various reasons, usually having to do with the watch list)
* Refunded (not the same as 'paid and refunded,' these are fully refunded badges that cannot check in)

We may want this as a plugin, but it's useful enough and hooks into a lot of other different features that I'm thinking of putting it in core. @EliAndrewC thoughts on that?

### PR DESCRIPTION
**Pull request by kitsuta**
Fixes #1222 and, in the process, attempts to address #199.

### CODE REVIEW

File: uber/static/theme/prereg.css [hunk 1]
```diff
@@ -66,8 +66,8 @@ body {
 }
 
 .prereg-price-notice {
-    color: red;
```
Comment (kitsuta) [inline uber/static/theme/prereg.css]: This is an unrelated fix; I found the price bump warnings to be a little too unreadable. :)

File: uber/models.py [hunk 2]
```diff
@@ -615,8 +615,8 @@ def change_badge(self, attendee, badge_type, badge_num=None):
             return 'Badge updated'
 
         def everyone(self):
```
Comment (kitsuta) [inline uber/models.py]: Because the idea of the badge statuses is that we only care about New and Complete badges, I changed the `everyone` variable here to reflect that.

File: uber/models.py [hunk 3]
```diff
@@ -1187,6 +1197,10 @@ def is_group_leader(self):
         return self.group and self.id == self.group.leader_id
 
     @property
+    def can_check_in(self):
```
Comment (kitsuta) [inline uber/models.py]: This isn't currently used, but will be used later to determine whether or not to display the 'check in' button.

File: uber/configspec.ini [hunk 4]
```diff
@@ -446,6 +446,13 @@ guest_badge     = string(default="Guest")
 one_day_badge   = string(default="One Day")
 __many__ = string
 
+[[badge_status]]
```
Comment (kitsuta) [inline uber/configspec.ini]: These are the statuses that every convention will definitely use.

I'm conflicted on whether or not to include a `printed_status` variable here - it was used a bunch for the logic in Anthrocon, but should it be included in core even though not every convention will need it as a status? Hm.

**Comment by kitsuta** [comment]
The badge status code has been well-tested in Anthrocon, so I'm not really worried about that. But what I'd like to discuss here is the database migration method.

Sqitch is a ORM-agnostic database migration software that uses native DDL to perform migrations. You can view a big tutorial here: https://github.com/theory/sqitch/blob/master/lib/sqitchtutorial.pod

In order to properly use and create migrations, users will have to install Sqitch itself, plus apgdiff (the latter is optional for production systems). I used cpan to install Sqitch and apt-get to install apgdiff.

After installation, you will have to fast-forward sqitch to master-base like so:
```
sudo su - postgres
cd /home/vagrant/uber/sideboard/plugins/uber/db
sqitch deploy db:pg:m13 master-base --log-only
```

To simply deploy migrations (such as the one included in this PR), use `sqitch deploy db:pg:m13`

Creating the migrations is a fair bit more complex, but generally works like this:
```
# First, edit models.py to reflect the changes you want to make, then reset your dev database to get the new model
sep reset_uber_db
sudo su - postgres
cd /home/vagrant/uber/sideboard/plugins/uber/db
sqitch add NAME-OF-FEATURE-BRANCH -n 'Migration message.'
pg_dump -s m13 > raw/NAME-OF-FEATURE-BRANCH.sql
apgdiff raw/master-current.sql raw/NAME-OF-FEATURE-BRANCH.sql
# Take the output and copy it into deploy/NAME-OF-FEATURE-BRANCH.sql
apgdiff raw/NAME-OF-FEATURE-BRANCH.sql raw/master-current.sql
# Take the output and copy it into revert/NAME-OF-FEATURE-BRANCH.sql
# Optionally, write a verification script in verify/NAME-OF-FEATURE-BRANCH.sql, like trying to select the column you just created
pg_dump -s m13 > raw/master-current.sql
exit
```

As you can see, it's a fair bit of commands, but it won't take too much work to completely automate them. The end result is that you have a raw/ folder that contains DDL to create any version of the database from scratch, _plus_ a set of migration scripts.

The big caveat here is that if you have a plugin that alters your model, you'll end up creating migrations that have extra garbage in them. The solution is, uh, don't do that. :'D

**Comment by kitsuta** [comment]
A few extra thoughts:

I tried to set up Alembic three times. We actually used it for Anthrocon, but we had some problems when my dev environment and the production environment got out of sync. More importantly, the --autogenerate feature is completely broken thanks to a very odd import feedback loop, so all migrations have to be written by hand in Alembic's special custom syntax.

Alembic could still be the better choice here, but:
* I'm not a fan of making developers learn yet another set of commands for database operations. Specifically, I'm worried about someone will write the migration in a way that doesn't match what sqlalchemy actually does when it inspects models.py, and it won't be caught until we have a new instance - which could be months after the migration is written.
* We may want to refactor the database or the ORM we use in the future, in which case we'd have to jump to another migration system.
* I've spent over a dozen hours trying to make it work, it's been a year since we said we wanted database migrations, and I'd rather have something inferior to Alembic than nothing at all.

I've just set up Sqitch today, and I'm pretty certain there's some configuration niceties that we can do cool stuff with. I know you can configure database targets, I'd like to puppetize that possibly. It also lets you set up users, which would be very nice but I'm not sure it's feasible with the vagrant setup.

Thanks to the way Sqitch works, you can have branches of database versions, so we could theoretically also create migrations inside of plugins. Again, though, we have to be careful not to mix the master branch database versions and the plugin database versions.

File: uber/models.py [hunk 5]
```diff
@@ -1090,6 +1091,15 @@ def _badge_adjustments(self):
                     self.badge_num = self.session.next_badge_num(self.badge_type, old_badge_num=0)
 
     @presave_adjustment
+    def _status_adjustments(self):
+        if self.badge_status == c.NEW_STATUS and not self.placeholder and self.first_name:
+            if (self.paid == c.HAS_PAID or self.paid == c.NEED_NOT_PAY) or (self.paid == c.PAID_BY_GROUP and not self.group.amount_unpaid):
```
Comment (EliAndrewC) [inline uber/models.py]: I would rewrite the first two statements to just say ``if self.badge_status in [c.HAS_PAID, c.NEED_NOT_PAY``.

Also, all of the parenthesis on this line are unnecessary, though I understand that some people prefer ``a or (b and c)`` since even though it's unnecessary it helps for people who don't always remember the operator precedence.

File: uber/models.py [hunk 6]
```diff
@@ -1090,6 +1091,15 @@ def _badge_adjustments(self):
...
+                self.badge_status = c.COMPLETED_STATUS
+        elif self.badge_status == c.INVALID_STATUS and self.admin_account:
```
Comment (EliAndrewC) [inline uber/models.py]: Is this ``Tracking`` call necessary here?  Shouldn't that happen automatically when the thing is deleted?

Or does that not happen automatically because we're already inside the on-update event?

File: uber/models.py (hunk 2 above)
Comment (EliAndrewC) [inline uber/models.py]: I'm not sure that ``everyone`` is even used anywhere since it doesn't have good performance, but I'm glad we're keeping it up to date so long as it exists.

File: uber/configspec.ini (hunk 4 above)
Comment (EliAndrewC) [inline uber/configspec.ini]: I'd leave it out of core if it's not something intrinsic.  As a general rule of thumb, if something doesn't have actual logic in core that refers to ``c.WHATEVER`` then it probably shouldn't be included in the core ``configspec.ini``.  I've gradually been deleting all of the things from ``configspec.ini`` which don't meet that criteria, although there are currently plenty of exceptions I haven't gotten to yet.

Side note: I'd personally much prefer ``c.NEW`` and ``c.COMPLETED`` to ``c.NEW_STATUS`` and ``c.COMPLETED_STATUS``.  It's not a problem if the same constant ends up in two different enums, so I don't think it's necessary to append ``_STATUS`` to the end of every one of these.

**Comment by EliAndrewC** [comment]
At a glance and from skimming through the docs, Squitch seems fine.

I'll need to give it a closer gander before I can offer much of an informed opinion, but since I'm not a huge fan of alembic anyway I have no real objection to switching to something else.  (We use alembic at my job and it works fine for us, but it's not nearly as nice as Django's built-in migration stuff and requires a lot more manual work and testing.)

One thing we'll need to think about is how to handle something like this:
- core RAMS has a set of migrations
- two different projects each add their own custom fields, so their stating points are different
- core RAMs adds a migration because we added a DB column or something
- each plugin adds a new table or column, so they have their own migrations
- core RAMs adds another migration because we add something else to core

In general, most migration tools assume that you have a non-branching sequence of migrations.  In this case we'd have migrations A, B, C, and D where one event wants A->B->D and another event wants A->C->D and that's not even accounting for how the fact that the starting point is different for each of them.

This seems solvable, but there'll probably be a lot of jumping through hoops.

**Comment by EliAndrewC** [comment]
Setting aside the DB migration stuff for the moment, let me make sure I understand the current implementation of badge statuses:
- at present, the status can be edited on the admin form
- the status is set automatically based on a few heuristics such as whether or not someone is paid
- in this current implementation, badge statuses and attendee deletion are orthogonal; an admin could at present either click the Delete button on an attendee or set their status to invalid

Assuming this has all been tested locally, I'm fine with merging it.

File: uber/models.py (hunk 2 above)
Comment (kitsuta) [inline uber/models.py]: Last I checked it was used in summary.py ... which explains why that page has so many load problems.

File: uber/configspec.ini (hunk 4 above)
Comment (kitsuta) [inline uber/configspec.ini]: It may not be a problem for the program itself, but if I wanted to do a search for everywhere I check for a "Completed" badge status then I'd definitely want it to be a unique name. That's why I named them like this, even though it's a bit more verbose.

File: uber/models.py (hunk 5 above)
Comment (kitsuta) [inline uber/models.py]: Whoops, totally missed the fact that those first two statements were inefficient.

File: uber/models.py (hunk 6 above)
Comment (kitsuta) [inline uber/models.py]: I... am not sure, actually.

**Comment by kitsuta** [comment]
I'm not sure what you mean by the third point, but just to clarify, badges can still be deleted-deleted. Otherwise that is all correct.

Much as it pains me to say (since we've already launched pre-reg and we'll have more badges to manually fix the longer we wait), let's wait to merge this until things have calmed down a bit. Someone will need to actually migrate the production database, after all.
